src/components/Liveness/Recorder.js

```diff
@@ -0,0 +1,196 @@
+  }
+
+  handleInactivityTimeout = () => {
+    this.setState({ hasBecomeInactive: true })
+  }
+
+  handleRecordingTimeout = () => {
+    this.setState({ hasRecordingTakenTooLong: true })
+    this.stopRecording()
+  }
+
+  redoActionsFallback = (text: string) =>
+    <span onClick={this.props.onRedo}>{text}</span>
+
+  renderError = () => {
+    const { hasRecordingTakenTooLong, hasBecomeInactive } = this.state
+    const { i18n, trackScreen, renderFallback } = this.props
+    if (hasBecomeInactive) {
+      return (
+        <CameraError
+          error={inactiveError}
+          hasBackdrop
+          {...{i18n, trackScreen, renderFallback }}
+        />
+      )
+    }
+
+    if (hasRecordingTakenTooLong) {
+      return (
+        <CameraError
+          error={recordingTooLongError}
+          renderFallback={ this.redoActionsFallback }
+          hasBackdrop
+          {...{i18n, trackScreen}}
+        />
+      )
+    }
+
+    return null
+  }
+
+  render = () => {
+    const { i18n, challenges = [], useFullScreen } = this.props
+    const { isRecording, currentIndex, hasBecomeInactive, hasRecordingTakenTooLong } = this.state
+    const currentChallenge = challenges[currentIndex] || {}
+    const isLastChallenge = currentIndex === challenges.length - 1
+    const hasError = hasBecomeInactive || hasRecordingTakenTooLong
+
+    return (
+      <div>
+        <Camera
+          {...this.props}
+          webcamRef={ c => this.webcam = c }
+          renderTitle={ !isRecording ?
+            <Title title={i18n.t('capture.liveness.challenges.position_face')} isFullScreen /> :
+            null
+          }
+          renderError={ hasError ? this.renderError() : undefined }
+          video
+        >
+          <ToggleFullScreen {...{useFullScreen}} />
+          {
+            isRecording ?
+              <Timeout key="recording" seconds={ 20 } onTimeout={ this.handleRecordingTimeout } /> :
+              <Timeout key="notRecording" seconds={ 12 } onTimeout={ this.handleInactivityTimeout } />
+          }
+          <FaceOverlay isFullScreen isWithoutHole={ hasError || isRecording } />
+          <div className={style.caption}>
+          {
+            isRecording &&
+              <div>
+                <div className={style.recordingIndicator}>
+                  {i18n.t('capture.liveness.recording')}
+                </div>
+                <Challenge {...{i18n, ...currentChallenge }} />
+              </div>
+          }
+          </div>
+          <div className={style.actions}>
+            <div className={classNames(style.captureActionsHint, {
+              [style.recordAction]: !isRecording,
+            })}>
+              { isRecording ?
+                i18n.t(`capture.liveness.challenges.done_${ isLastChallenge ? 'stop' : 'next' }`) :
+                i18n.t('capture.liveness.press_record')
+              }
+            </div>
+            {
+              !isLastChallenge && isRecording ?
```

Comment:
There are a lot of conditions in this render. Could we group some of them in named functions so it's easier to read?